File: c++/100.cpp
// 100. Same Tree
// Given two binary trees, write a function to check if they are the same or not.
//
// Two binary trees are considered the same if they are structurally identical and the nodes have the same value.
//
// Example 1:
//
// Input:     1         1
//           / \       / \
//          2   3     2   3
//
//         [1,2,3],   [1,2,3]
//
// Output: true
// Example 2:
//
// Input:     1         1
//           /           \
//          2             2
//
//         [1,2],     [1,null,2]
//
// Output: false
// Example 3:
//
// Input:     1         1
//           / \       / \
//          2   1     1   2
//
//         [1,2,1],   [1,1,2]
//
// Output: false

// 0 ms, faster than 100.00%
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
  bool isSameTree(TreeNode* p, TreeNode* q) {
    if (p == NULL) {
      if (q == NULL) return true;
      else
        return false;
    }
    if (p != NULL && q == NULL) return false;
    if (p->val != q->val) {
      return false;
    }
    if (!isSameTree(p->left, q->left)) return false;
    if(!isSameTree(p->right, q->right)) return false;
      return true;
  }
};

File: go/100.go
// 100. Same Tree

// Given the roots of two binary trees p and q, write a function to check if they are the same or not.

// Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

 

// Example 1:


// Input: p = [1,2,3], q = [1,2,3]
// Output: true
// Example 2:


// Input: p = [1,2], q = [1,null,2]
// Output: false
// Example 3:


// Input: p = [1,2,1], q = [1,1,2]
// Output: false
 

// Constraints:

// The number of nodes in both trees is in the range [0, 100].
// -104 <= Node.val <= 104

// Runtime 0 ms Beats 100.00%
// Memory 4.09 MB Beats 99.40%
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func isSameTree(p *TreeNode, q *TreeNode) bool {
    if p == nil && q == nil {
        return true
    }
    if (p != nil && q == nil) || (p == nil && q != nil) || (p != nil && q != nil && p.Val != q.Val) {
        return false
    }
    isSameLeft := isSameTree(p.Left, q.Left)
    isSameRight := isSameTree(p.Right, q.Right)
    return isSameLeft && isSameRight
}